text/0141-linking-traces.md

```diff
@@ -0,0 +1,367 @@
+
+#### Usage Example
+
+Adding span links should be possible at span start time, as well as when holding a reference to the span:
+
+```typescript
+// 1st trace starts
+const pageloadSpan = startInactiveSpan(...)
+
+// 2nd trace starts
+const navigation1Span = startInactiveSpan({
+  name: '/users', 
+  links: [{
+    context: pageloadSpan.spanContext(), 
+    attributes: {
+      'sentry.link.type': 'previous_trace'
+    }
+  }]
+});
+
+// 3rd trace starts
+const navigation2Span = startSpan({name: '/users/:id'}, (span) => {
+  span.addLink({
+    context: navigation1Span.spanContext(), 
+    attributes: {
+      'sentry.link.type': 'previous_trace'
+    }
+  })
+})
+```
+
+In this example, by adding span links, we can link from the last navigation trace all the way back to the initial pageload trace. By passing the `'sentry.link.type': 'previous_trace'` attribute, we can identify the link as a previous trace link in Sentry and display the spans accordingly. 
+
+#### SDKs relying on Transactions
+
+Adding span links should behave identically regardless of links being added to a root/segment (transaction) or child spans. For SDKs still having public APIs around transactions, their respective `Transaction` interface and `startTransaction` function(s) should support the same APIs. 
+
+#### Storing links and previous traces
+
+SDKs are free to chose whatever storage mechanism makes sense to store span links themselves as well as to hold references to a previous span. This is considered an implementation detail and out of scope for this RFC.
+
+### Envelope Item Payload Changes
+
+Span trees are serialized to transaction event envelopes in all Sentry SDKs. Therefore, the envelope item needs to accommodate span links in its payload.
+
+We propose to store span links in the `trace` context if the root span has links, similarly to how we serialize root span attributes today:
+
+```typescript
+// event envelope item
+{
+  type: "transaction";
+  transaction: string;
+  contexts: {
+    trace: {
+      span_id: string;
+      parent_span_id: string;
+      trace_id: string;
+      // new field for links:
+      links?: Array<{
+        "span_id": string,
+        "trace_id": string, 
+        attributes: Record<string, AttributeValue>,
+      }>
+      // ...
+    }
+  }
+  // ...
+}
+```
+
+For links stored in child spans, SDKs should serialize them to `spans[i].links`:
+
+```typescript
+// event envelope item
+{
+  type: "transaction";
+  transaction: string;
+  spans: Array<{
+    span_id: string;
+    parent_span_id: string;
+    trace_id: string;
+    // new field for links:
+    links?: Array<{
+      "span_id": string,
+      "trace_id": string, 
+      attributes: Record<string, AttributeValue>,
+    }>
+    // ...
+  }>
+  // ...
+}
+```
+
```

Comment:
#### Pros

1. Allows adding multiple span links to one span
2. Allows linking to other data types, such as errors, session replays, user feedback, sessions, and future data types.
3. Extendable: We can add new types of links without having to change the API.
4. Compatible with OTEL
5. Is compatible with traces, transactions, and spans. We can link everything that has a trace, such as errors.
6. Should be compatible with a potential future session ID. If we at some point add a session ID or use sessions as a session ID to link all data to one session, we should only have to add a `'sentry.link.type':  'session'`.

Reply:
> Allows linking to other data types, such as errors, session replays, user feedback, sessions, and future data types.

Sorry, I might be missing something but how does this proposal allow you to do that?
Typically everything goes the opposite direction (everything else has traceId and spanId properties) - what would be the use case for linking in the opposite direction?

> Should be compatible with a potential future session ID. If we at some point add a session ID or use sessions as a session ID to link all data to one session, we should only have to add a 'sentry.link.type':  'session'.

FWIW the de facto standard way to handle sessions would be with a `session.id` attribute - not sure what the advantages of a link would be over that? - but that's all down the road if ever.

Reply:
> what would be the use case for linking in the opposite direction?

You're right. Most of the time, all the data types link to a trace. Linking a span or multiple spans to an error could be one use case. I guess we should change the pro to:
2. Allows linking to other data types, such as errors and future data types.  

> FWIW the de facto standard way to handle sessions would be with a session.id attribute - not sure what the advantages of a link would be over that? - but that's all down the road if ever.

I'm not saying we have to do it as I described it; we could only use the same approach for sessions if we needed to. Yes, that's down the road.

Reply:
I would argue that we should focus on span links as a mechanism to link **only to other spans**. Linking to other events is out of the otel spec as far as I can tell and something that would require some "API bending". This we can't do easily with POTEL SDKs so let's not focus on this too much. Does that sound reasonable to you?

Reply:
Yes, that makes sense. Please put it in the RFC, and let's not add this as a pro.

Reply:
I added a note about this in the RFC.